diff --git a/doc/_U_dyn_register.man b/doc/_U_dyn_register.man
index 107e5fd..83d0339 100644
--- a/doc/_U_dyn_register.man
+++ b/doc/_U_dyn_register.man
@@ -1,5 +1,5 @@
 '\" t
-.\" Manual page created with latex2man on Thu Aug 16 09:44:45 MDT 2007
+.\" Manual page created with latex2man on Fri Apr 18 13:21:42 CEST 2014
 .\" NOTE: This file is generated, DO NOT EDIT.
 .de Vb
 .ft CW
@@ -10,7 +10,7 @@
 
 .fi
 ..
-.TH "\\_U\\_DYN\\_REGISTER" "3" "16 August 2007" "Programming Library " "Programming Library "
+.TH "\\_U\\_DYN\\_REGISTER" "3" "18 April 2014" "Programming Library " "Programming Library "
 .SH NAME
 _U_dyn_register
 \-\- register unwind\-info for dynamically generated code 
diff --git a/include/config.h b/include/config.h
deleted file mode 100644
index 1447c58..0000000
--- a/include/config.h
+++ /dev/null
@@ -1,256 +0,0 @@
-/* include/config.h.  Generated from config.h.in by configure.  */
-/* include/config.h.in.  Generated from configure.ac by autoheader.  */
-
-/* Block signals before mutex operations */
-#define CONFIG_BLOCK_SIGNALS /**/
-
-/* Enable Debug Frame */
-#if defined(__arm__)
-#define CONFIG_DEBUG_FRAME /**/
-#endif
-
-/* Support for Microsoft ABI extensions */
-/* #undef CONFIG_MSABI_SUPPORT */
-
-/* Define to 1 if you want every memory access validated */
-#define CONSERVATIVE_CHECKS 1
-
-/* Allocate large structures rather than place them on the stack. */
-#define CONSERVE_STACK /**/
-
-/* Define to 1 if you have the <asm/ptrace_offsets.h> header file. */
-/* #undef HAVE_ASM_PTRACE_OFFSETS_H */
-
-/* Define to 1 if you have the <atomic_ops.h> header file. */
-/* #undef HAVE_ATOMIC_OPS_H */
-
-/* Define to 1 if you have the <byteswap.h> header file. */
-#define HAVE_BYTESWAP_H 1
-
-/* Define to 1 if you have the declaration of `PTRACE_CONT', and to 0 if you
-   don't. */
-#define HAVE_DECL_PTRACE_CONT 1
-
-/* Define to 1 if you have the declaration of `PTRACE_POKEDATA', and to 0 if
-   you don't. */
-#define HAVE_DECL_PTRACE_POKEDATA 1
-
-/* Define to 1 if you have the declaration of `PTRACE_POKEUSER', and to 0 if
-   you don't. */
-#if defined(__aarch64__) || defined(__mips__)
-#define HAVE_DECL_PTRACE_POKEUSER 0
-#else
-#define HAVE_DECL_PTRACE_POKEUSER 1
-#endif
-
-/* Define to 1 if you have the declaration of `PTRACE_SINGLESTEP', and to 0 if
-   you don't. */
-#define HAVE_DECL_PTRACE_SINGLESTEP 1
-
-/* Define to 1 if you have the declaration of `PTRACE_SYSCALL', and to 0 if
-   you don't. */
-#define HAVE_DECL_PTRACE_SYSCALL 1
-
-/* Define to 1 if you have the declaration of `PTRACE_TRACEME', and to 0 if
-   you don't. */
-#define HAVE_DECL_PTRACE_TRACEME 1
-
-/* Define to 1 if you have the declaration of `PT_CONTINUE', and to 0 if you
-   don't. */
-#define HAVE_DECL_PT_CONTINUE 0
-
-/* Define to 1 if you have the declaration of `PT_GETFPREGS', and to 0 if you
-   don't. */
-#define HAVE_DECL_PT_GETFPREGS 0
-
-/* Define to 1 if you have the declaration of `PT_GETREGS', and to 0 if you
-   don't. */
-#if defined(__mips__)
-#define HAVE_DECL_PT_GETREGS 1
-#else
-#define HAVE_DECL_PT_GETREGS 0
-#endif
-
-/* Define to 1 if you have the declaration of `PT_GETREGSET', and to 0 if you
-   don't. */
-#if defined(__aarch64__)
-#define HAVE_DECL_PT_GETREGSET 1
-#else
-#define HAVE_DECL_PT_GETREGSET 0
-#endif
-
-/* Define to 1 if you have the declaration of `PT_IO', and to 0 if you don't.
-   */
-#define HAVE_DECL_PT_IO 0
-
-/* Define to 1 if you have the declaration of `PT_STEP', and to 0 if you
-   don't. */
-#define HAVE_DECL_PT_STEP 0
-
-/* Define to 1 if you have the declaration of `PT_SYSCALL', and to 0 if you
-   don't. */
-#define HAVE_DECL_PT_SYSCALL 0
-
-/* Define to 1 if you have the declaration of `PT_TRACE_ME', and to 0 if you
-   don't. */
-#define HAVE_DECL_PT_TRACE_ME 0
-
-/* Define to 1 if you have the <dlfcn.h> header file. */
-#define HAVE_DLFCN_H 1
-
-/* Define to 1 if you have the `dlmodinfo' function. */
-#define HAVE_DLMODINFO 1
-
-/* Define to 1 if you have the `dl_iterate_phdr' function. */
-#define HAVE_DL_ITERATE_PHDR 1
-
-/* Define to 1 if you have the `dl_phdr_removals_counter' function. */
-#define HAVE_DL_PHDR_REMOVALS_COUNTER 1
-
-/* Define to 1 if you have the <elf.h> header file. */
-#define HAVE_ELF_H 1
-
-/* Define to 1 if you have the <endian.h> header file. */
-//#define HAVE_ENDIAN_H 1
-
-/* Define to 1 if you have the <execinfo.h> header file. */
-/* #undef HAVE_EXECINFO_H */
-
-/* Define to 1 if you have the `getunwind' function. */
-#define HAVE_GETUNWIND 1
-
-/* Define to 1 if you have the <ia64intrin.h> header file. */
-/* #undef HAVE_IA64INTRIN_H */
-
-/* Define to 1 if you have the <inttypes.h> header file. */
-#define HAVE_INTTYPES_H 1
-
-/* Define to 1 if you have the `uca' library (-luca). */
-/* #undef HAVE_LIBUCA */
-
-/* Define to 1 if you have the <link.h> header file. */
-#define HAVE_LINK_H 1
-
-/* Define if you have liblzma */
-/* #undef HAVE_LZMA */
-
-/* Define to 1 if you have the <memory.h> header file. */
-#define HAVE_MEMORY_H 1
-
-/* Define to 1 if you have the `mincore' function. */
-#define HAVE_MINCORE 1
-
-/* Define to 1 if you have the <signal.h> header file. */
-#define HAVE_SIGNAL_H 1
-
-/* Define to 1 if you have the <stdint.h> header file. */
-#define HAVE_STDINT_H 1
-
-/* Define to 1 if you have the <stdlib.h> header file. */
-#define HAVE_STDLIB_H 1
-
-/* Define to 1 if you have the <strings.h> header file. */
-#define HAVE_STRINGS_H 1
-
-/* Define to 1 if you have the <string.h> header file. */
-#define HAVE_STRING_H 1
-
-/* Define to 1 if `dlpi_subs' is a member of `struct dl_phdr_info'. */
-/* #undef HAVE_STRUCT_DL_PHDR_INFO_DLPI_SUBS */
-
-/* Define to 1 if the system has the type `struct elf_prstatus'. */
-/* #undef HAVE_STRUCT_ELF_PRSTATUS */
-
-/* Define to 1 if the system has the type `struct prstatus'. */
-/* #undef HAVE_STRUCT_PRSTATUS */
-
-/* Defined if __sync atomics are available */
-#define HAVE_SYNC_ATOMICS 1
-
-/* Define to 1 if you have the <sys/elf.h> header file. */
-/* #undef HAVE_SYS_ELF_H */
-
-/* Define to 1 if you have the <sys/endian.h> header file. */
-#define HAVE_SYS_ENDIAN_H 1
-
-/* Define to 1 if you have the <sys/link.h> header file. */
-/* #undef HAVE_SYS_LINK_H */
-
-/* Define to 1 if you have the <sys/procfs.h> header file. */
-/* #undef HAVE_SYS_PROCFS_H */
-
-/* Define to 1 if you have the <sys/ptrace.h> header file. */
-#define HAVE_SYS_PTRACE_H 1
-
-/* Define to 1 if you have the <sys/stat.h> header file. */
-#define HAVE_SYS_STAT_H 1
-
-/* Define to 1 if you have the <sys/types.h> header file. */
-#define HAVE_SYS_TYPES_H 1
-
-/* Define to 1 if you have the <sys/uc_access.h> header file. */
-/* #undef HAVE_SYS_UC_ACCESS_H */
-
-/* Define to 1 if you have the `ttrace' function. */
-/* #undef HAVE_TTRACE */
-
-/* Define to 1 if you have the <unistd.h> header file. */
-#define HAVE_UNISTD_H 1
-
-/* Defined if __builtin_unreachable() is available */
-#define HAVE__BUILTIN_UNREACHABLE 1
-
-/* Defined if __builtin___clear_cache() is available */
-#define HAVE__BUILTIN___CLEAR_CACHE 1
-
-/* Define to 1 if __thread keyword is supported by the C compiler. */
-#define HAVE___THREAD 1
-
-/* Define to the sub-directory in which libtool stores uninstalled libraries.
-   */
-#define LT_OBJDIR ".libs/"
-
-/* Define to 1 if your C compiler doesn't accept -c and -o together. */
-/* #undef NO_MINUS_C_MINUS_O */
-
-/* Name of package */
-#define PACKAGE "libunwind"
-
-/* Define to the address where bug reports for this package should be sent. */
-#define PACKAGE_BUGREPORT "libunwind-devel@nongnu.org"
-
-/* Define to the full name of this package. */
-#define PACKAGE_NAME "libunwind"
-
-/* Define to the full name and version of this package. */
-#define PACKAGE_STRING "libunwind 1.1"
-
-/* Define to the one symbol short name of this package. */
-#define PACKAGE_TARNAME "libunwind"
-
-/* Define to the home page for this package. */
-#define PACKAGE_URL ""
-
-/* Define to the version of this package. */
-#define PACKAGE_VERSION "1.1"
-
-/* The size of `off_t', as computed by sizeof. */
-#define SIZEOF_OFF_T 4
-
-/* Define to 1 if you have the ANSI C header files. */
-#define STDC_HEADERS 1
-
-/* Version number of package */
-#define VERSION "1.1"
-
-/* Define to empty if `const' does not conform to ANSI C. */
-/* #undef const */
-
-/* Define to `__inline__' or `__inline' if that's what the C compiler
-   calls it, or to nothing if 'inline' is not supported under any name.  */
-#ifndef __cplusplus
-/* #undef inline */
-#endif
-
-/* Define to `unsigned int' if <sys/types.h> does not define. */
-/* #undef size_t */
diff --git a/include/libunwind-common.h.in b/include/libunwind-common.h.in
index fa753ba..a1dadec 100644
--- a/include/libunwind-common.h.in
+++ b/include/libunwind-common.h.in
@@ -23,6 +23,10 @@ LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 
+/* ANDROID support update. */
+#include <sys/types.h>
+/* End of ANDROID update. */
+
 #define UNW_VERSION_MAJOR	@PKG_MAJOR@
 #define UNW_VERSION_MINOR	@PKG_MINOR@
 #define UNW_VERSION_EXTRA	@PKG_EXTRA@
@@ -204,6 +208,24 @@ typedef struct unw_save_loc
   }
 unw_save_loc_t;
 
+/* ANDROID support update. */
+typedef struct unw_map_cursor
+  {
+    void *map_list;
+    void *cur_map;
+  }
+unw_map_cursor_t;
+
+typedef struct unw_map
+  {
+    unw_word_t start;
+    unw_word_t end;
+    char *path;
+    int flags;
+  }
+unw_map_t;
+/* End of ANDROID update. */
+
 /* These routines work both for local and remote unwinding.  */
 
 #define unw_local_addr_space	UNW_OBJ(local_addr_space)
@@ -224,6 +246,7 @@ unw_save_loc_t;
 #define unw_is_signal_frame	UNW_OBJ(is_signal_frame)
 #define unw_handle_signal_frame	UNW_OBJ(handle_signal_frame)
 #define unw_get_proc_name	UNW_OBJ(get_proc_name)
+#define unw_get_proc_name_by_ip	UNW_OBJ(get_proc_name_by_ip)
 #define unw_set_caching_policy	UNW_OBJ(set_caching_policy)
 #define unw_regname		UNW_ARCH_OBJ(regname)
 #define unw_flush_cache		UNW_ARCH_OBJ(flush_cache)
@@ -251,7 +274,23 @@ extern int unw_get_save_loc (unw_cursor_t *, int, unw_save_loc_t *);
 extern int unw_is_signal_frame (unw_cursor_t *);
 extern int unw_handle_signal_frame (unw_cursor_t *);
 extern int unw_get_proc_name (unw_cursor_t *, char *, size_t, unw_word_t *);
+extern int unw_get_proc_name_by_ip (unw_addr_space_t, unw_word_t, char *,
+				    size_t, unw_word_t *, void *);
 extern const char *unw_strerror (int);
 extern int unw_backtrace (void **, int);
 
+/* ANDROID support update. */
+extern int unw_map_local_cursor_valid (unw_map_cursor_t *);
+extern void unw_map_local_cursor_get (unw_map_cursor_t *);
+extern int unw_map_local_cursor_get_next (unw_map_cursor_t *, unw_map_t *);
+extern int unw_map_local_create (void);
+extern void unw_map_local_destroy (void);
+extern void unw_map_set (unw_addr_space_t, unw_map_cursor_t *);
+extern void unw_map_cursor_reset (unw_map_cursor_t *);
+extern void unw_map_cursor_clear (unw_map_cursor_t *);
+extern int unw_map_cursor_create (unw_map_cursor_t *, pid_t);
+extern void unw_map_cursor_destroy (unw_map_cursor_t *);
+extern int unw_map_cursor_get_next (unw_map_cursor_t *, unw_map_t *);
+/* End of ANDROID update. */
+
 extern unw_addr_space_t unw_local_addr_space;
diff --git a/src/Makefile.am b/src/Makefile.am
index cf137c0..e0cf481 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -98,7 +98,7 @@ libunwind_la_LIBADD =
 # libraries:
 libunwind_la_SOURCES_common =					\
 	$(libunwind_la_SOURCES_os)				\
-	mi/init.c mi/flush_cache.c mi/mempool.c mi/strerror.c
+	mi/init.c mi/flush_cache.c mi/mempool.c mi/strerror.c mi/Lmap.c mi/map.c os-common.c Los-common.c
 
 # List of arch-independent files needed by generic library (libunwind-$ARCH):
 libunwind_la_SOURCES_generic =						\
