diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..a67f5bd
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,22 @@
+project(nativehelper)
+cmake_minimum_required(VERSION 2.8)
+
+include(/usr/share/android/cmake/AndroidConfig.cmake)
+find_package(Android REQUIRED )
+
+include_directories ("${PROJECT_SOURCE_DIR}/include/nativehelper" ${ANDROID_INCLUDE_DIRS})
+
+set(SRC_LIST JNIHelp.cpp 
+    JniConstants.cpp
+    toStringArray.cpp
+    JniInvocation.cpp)
+
+add_library(${PROJECT_NAME} STATIC ${SRC_LIST})
+SET_TARGET_PROPERTIES(${PROJECT_NAME} PROPERTIES COMPILE_FLAGS "-fPIC")
+target_link_libraries(${PROJECT_NAME} ${ANDROID_LIBRARIES})
+
+
+install(TARGETS ${PROJECT_NAME} DESTINATION ${INSTALL_LIB_DIR})
+install(DIRECTORY ${PROJECT_SOURCE_DIR}/include/nativehelper DESTINATION ${INSTALL_INCLUDE_DIR} FILES_MATCHING PATTERN "*.h")
+
+
diff --git a/include/nativehelper/ScopedFd.h b/include/nativehelper/ScopedFd.h
index 41ca7f9..246bede 100644
--- a/include/nativehelper/ScopedFd.h
+++ b/include/nativehelper/ScopedFd.h
@@ -18,6 +18,7 @@
 #define SCOPED_FD_H_included
 
 #include <unistd.h>
+#include <errno.h>
 #include "JNIHelp.h"  // for TEMP_FAILURE_RETRY
 
 // A smart pointer that closes the given fd on going out of scope.
diff --git a/include/nativehelper/UniquePtr.h b/include/nativehelper/UniquePtr.h
index 50f75b2..de78b47 100644
--- a/include/nativehelper/UniquePtr.h
+++ b/include/nativehelper/UniquePtr.h
@@ -18,12 +18,7 @@
 #define UNIQUE_PTR_H_included
 
 #include <cstdlib> // For NULL.
-
-// This is a fake declaration of std::swap to avoid including <algorithm>
-namespace std {
-template <class T> void swap(T&, T&);
-}
-
+#include <algorithm> 
 // Default deleter for pointer types.
 template <typename T>
 struct DefaultDelete {
